README.md
# Umoria COLOR

_The Dungeons of Moria_ is a single player dungeon simulation originally
written by Robert Alan Koeneke, with its first public release in 1983.
The game was originally developed using VMS Pascal before being ported to the
C language by James E. Wilson in 1988, and released a _Umoria_.

Moria/Umoria has had many variants over the years, with [_Angband_](http://rephial.org/)
being the most well known. Umoria was also an inspiration for one the most
commercially successful action roguelike games, _Diablo_!

Supported Platforms:

  - Windows
  - macOS
  - Linux (Ubuntu/Debian)

Compiling and limited testing has been done for other Linux based system
including NetBSD 8.1 and Fedora 32.


## Umoria COLOR

Colour-Umoria was a patch originally written in 1993 by Edouard Poor. In 2020
it was ported by Andrew Weber as a fork of the modernized Umoria restoration
release.

* The correct spelling of "color" is now used 🇺🇸🦅🗽
* Many colors and color assignments were adjusted
* Roguelike controls are now disabled by default to encourage new players to try the game out
* Walking into a door will attempt to open it automatically without having to press the `'o'` key
* Fire, lightning, and glowing effects were added

## Umoria 5.7.x releases

The main focus of the `5.7.0` release was to provide support for the three
main operating systems: Windows, macOS, and Linux. Support for all other
outdated computer systems such as MS DOS, "Classic" Mac OS (pre OSX), Amiga,
and Atari ST was removed.

_Note: there have been no intentional game play changes in the 5.7.x releases._

Since the initial 5.7 release, a great deal of _code restoration_ has been
undertaken in the hope of aiding future development of the game. Some examples
of the work done include reformatting the source code with the help of
`clang-tidy` and `clang-format`, modernizing the code to use standard C types,
breaking apart most large functions (many of which had hundreds of lines of code)
into smaller, easier to read functions, and fixing all compiler warnings when
compiling against recent versions of GCC and Clang.

Full details of all changes can be found in the [CHANGELOG](CHANGELOG.md), and
by browsing the commit history.

Due to its lack of Windows and macOS support Moria was inaccessible to many
people. Hopefully these changes will give many more people a chance to play
this classic roguelike game.


## Notes on Compiling Umoria

Umoria has been tested against GCC (`10` and `11`) and with `ncurses 6.x`,
although recent earlier versions should also work fine.

You will need these as well as `CMake` and the C++ build tools for your system.


### macOS and Linux

Change to the `umoria` game directory and enter the following commands at the
terminal:

    $ mkdir build && cd build
    $ cmake ..
    $ make

NOTE: use `make -j $(nproc)` to speed up compilation on Linux.

An `umoria` directory will be created in the current directory containing the
game binary and data files, which can then be moved to any other location, such
as the `home` directory.


### Windows

MinGW is used to provide GCC and GNU Binutils for compiling on the Windows platform.
The easiest solution to get set up is to use the [MSYS2 Installer](http://msys2.github.io/).
Once installed, `pacman` can be used to install `GCC`, `ncurses`, and the
`make`/`cmake` build tools.

At present an environment variable for the MinGW system being compiled on will
need to be specified. This will be either `mingw64` or `mingw32`.

At the command prompt type the following, being sure to add the correct label
to `MINGW=`:

    $ MINGW=mingw64 cmake .
    $ make

To perform an out-of-source build, type the following:

    $ mkdir build
    $ cd build
    $ MINGW=mingw64 cmake ..
    $ make

As with the macOS/Linux builds, all files will be installed into an `umoria` directory.


## Historical Documents

Most of the original document files included in the Umoria 5.6 sources have
been placed in the [historical](historical) directory. You will even find the
old CHANGELOG, which tracks all code changes made between versions 4.81 and
5.5.2 (1987-2008). If you'd like to learn more on the development history of
Umoria, these can make for interesting reading.

There is also the original Moria Manual and FAQ. Although these are a little
outdated now they are certainly worth reading as they contain a lot of
interesting and useful information.


## Code of Conduct and Contributions

See here for details on our [Code of Conduct](CODE_OF_CONDUCT.md).

For details on how to contribute to the Umoria project, please read our
[contributing](CONTRIBUTING.md) guide.


## License Information

Umoria is released under the [GNU General Public License v3.0](LICENSE).

In 2007 Ben Asselstine and Ben Shadwick started the
[_free-moria_](http://free-moria.sourceforge.net/) project to re-license
UMoria 5.5.2 under GPL-2 by obtaining permission from all the contributing
authors. A year later they succeeded in their goal and in late 2008 official
maintainer David Grabiner released Umoria 5.6 under a GPL-3.0-or-later license.


CHANGELOG.md
# Umoria CHANGELOG

Tracking all changes since the 5.6.0 release.


## HEAD

- Rebased from upstream
- Revert to Moria keys as default.
- Minor configuration update to support compiling on Visual Studio 2019.


## 5.8.1 (2020-11-19)

- Bugfix displaying current experience color
- Option to enable/disable colors, with fallback for terminals without colors


## 5.8.0 (2020-11-18)

- Colors! Adapted from Colour-Umoria by Edouard Poor. Many colors and color assignments were adjusted
- Disable roguelike keys by default
- Walking into a door will attempt to open it automatically without having to press the `'o'` key
- Fire, lightning, and glowing effects were added


## 5.7.13 (2020-08-22)

- Bugfix in `getAndCastMagicSpell()`: losing mana when cancelling a mage spell


## 5.7.12 (2020-05-28)

- Bugfix in `pray()`: should loose mana when failing to recite a prayer.
- Add `Press ? for help` to the message bar on game start/load.
- Use roguelike keys by default, and remove setting option from the CLI.
- Player (`row`,`col`) and Monster (`y`,`x`) now use `Coord_t` for their positions.
- Use `Coord_t` in all functions that used `y, x` coordinates.
- Reorganise some game arrays and variables, placing them on an object:
    * put `inventory` onto `Player_t`
    * put `treasure_list` and `current_treasure_id` onto `Game_t`
- Type changes:
    * `rcmove` variable in `memoryUpdateRecall()` signature is now an `uint32_t` like everywhere else.
    * Change `store_buy` function signature type to use `uint8_t`, and return a `bool`.
    * Various other types changes.
- Typo fixes: `item_sub_catory`, `current_askin_price`.
- Various `clang-format` and `clang-tidy` updates.
- Move the Manual and FAQ to the `historical` directory and remove from release.
  It seems that these are very much out of date - possibly for any 5.x version - so
  it makes sense to move them back in with the rest of the historical documents.
- Various compiler fixes.


## 5.7.11 (2020-02-26)

- Rename several `ui_inventory.cpp` functions to avoid name clashes (bugfix).
- Player ToHit/Armor/Damage Adjustment functions now return `int16` types.
- Minor style changes.
- Various `CMakeLists.txt` updates:
    * Allow out-of-source builds.
    * Use `configure_file` to set variables in `splash.txt` and `versions.txt`.
    * Remove unneeded `make install`.
    * Fix for finding ncurses on Linux/macOS.
    * GCC 8 support.
- Update `AUTHORS`: add more known features from `-JWT-`.


## 5.7.10 (2018-02-18)

### Bug Fixes

- `xor_byte` decryption was not being performed correctly on `score.dat` files.
  Introduced with commit: 676cdfed6c274279fa889c079e84788adc954cac (`readHighScore()` function).
  Note: `game.sav` files were **not affected** by this bug.


### Code

- Delete `constant.h`, moving many constants into `config.h` and the rest into
  their related headers. Things are looking pretty ugly at the moment but
  perhaps it's a better starting point for further refactoring.
- Change `config.h` to namespace'd constants - perhaps not a good approach but let's see!


## 5.7.9 (2018-01-20)

- Add AUTHORS file containing all known author information.
  This removes contributors section from `versions.txt`.

### Bug Fixes

- Kill experience points now calculated correctly.
  When extracting a method the wrong `int` type was used when calculating the
  `creature.kill_exp_value * creature.level`.
  This bug was introduced in Umoria `5.7.3` with the commit: ccfa74783ad67eb3276ff3eca0f2509599012d33

### Code

Continuing the process of moving related functions to the same file, plus other
changes. Highlights:

- `types.h` now has just the two core `vtype_t` and `obj_desc_t` types.
- The numbered `misc` and `moria` files no longer exist!
- Moved `sets.cpp` functions elsewhere, allowing for most to become `static`.
- `externs.h` is now empty, so deleted!
- Add consistent `const` in various places.
- Use more `Coord_t` in Dungeon related functions.


## 5.7.8 (2017-12-24)

- Improve _Wizard Mode_ help files.
- Easier item creation in _Wizard Mode_ (inspired by Bmoria).
- Change `[Press any key to continue.]` to `[ press any key to continue ]`,
  and any variations, as it looks clearer.

### Code

There are two main areas of focus for this release. The first was to create
more objects to move the numerous global variables on to. E.g. `Dungeon_t`.

The second area of focus has been to start grouping related functions and
variables in the same file. Ex. most player functions are moved to `player.cpp`,
or a new `player_xxx.cpp` file has been created (run, stats, tunnel, etc.).
The LOS and look functions are located in `los.cpp`.

Grouping globals and functions together like this should make their usage and
their relationships more obvious.

These locations are by no means final, but are a useful first pass.

- Grouping related functions together in the same file.
- Move more Player globals to the `Player_t` struct.
- Create a `Dungeon_t` and put all dungeon related globals here.
- Create a `Dice_t` instead of using an array.
- Started replacing `char *` with `std::string`.
- Simpler display of _death_ screens (using `death_tomb.txt`, `death_royal.txt` files).


## 5.7.7 (2017-11-07)

### Bug Fixes

- Game loading fix where shops data was being read twice, and the second time
  resulted in incorrect data. Oops! Broken during save game refactor:
  https://github.com/dungeons-of-moria/umoria/commit/ce2c756


## 5.7.6 (2017-11-05)

### Bug Fixes

- When compiling on Windows/MinGW, the OS was not being detected correctly.
  https://github.com/dungeons-of-moria/umoria/commit/3811bcd
- Now load cave tiles correctly (the `lit` status values), as a previous
  refactoring broke loading of pre-5.7 saves.
  https://github.com/dungeons-of-moria/umoria/commit/219f350

### Code

- _Pass by Value_ and _Pass by Reference_ major refactoring.
- Replace "magic numbers" with ENUMs, in `staffs.cpp`.

src/color.h
// Copyright (c) 1981-86 Robert A. Koeneke
// Copyright (c) 1987-94 James E. Wilson
//
// This work is free software released under the GNU General Public License
// version 2.0, and comes with ABSOLUTELY NO WARRANTY.
//
// See LICENSE and AUTHORS for more information.

#pragma once

// Types used by the color routines
typedef struct {
  unsigned char L; // the local value for the particular color
  unsigned char R; // The full RGB specification for the color
  unsigned char G; // for those players with 24 bit displays
  unsigned char B; // (lucky so-and-so's...)
} Color_t;

/* Greys */
#define Color_White            0
#define Color_Light_Grey_High  1
#define Color_Light_Grey_Low   2
#define Color_Medium_Grey_High 3
#define Color_Medium_Grey_Low  4
#define Color_Dark_Grey_High   5
#define Color_Dark_Grey_Low    6
#define Color_Black            7

/* Standard */
#define Color_Light_Red        8
#define Color_Light_Green      9
#define Color_Light_Yellow     10
#define Color_Light_Blue       11
#define Color_Light_Orange     12
#define Color_Light_Purple     13
#define Color_Red              14
#define Color_Green            15
#define Color_Yellow           16
#define Color_Blue             17
#define Color_Orange           18
#define Color_Purple           19   /* Now a Major Motion Picture */
#define Color_Dark_Red         20
#define Color_Dark_Green       21
#define Color_Dark_Yellow      22
#define Color_Dark_Blue        23
#define Color_Dark_Orange      24
#define Color_Dark_Purple      25
#define Color_Cream            26

/* Amulets */
#define Color_Amber            27
#define Color_Driftwood        28
#define Color_Coral            29
#define Color_Agate            30
#define Color_Ivory            31
#define Color_Obsidian         32
#define Color_Bone             33
#define Color_Brass            34
#define Color_Bronze           35
#define Color_Pewter           36
#define Color_Tortoise_Shell   37

/* Metals */
#define Color_Aluminum         38
#define Color_Cast_Iron        39
#define Color_Chromium         40
#define Color_Copper           41
#define Color_Gold             42
#define Color_Iron             43
#define Color_Magnesium        44
#define Color_Molybdenum       45
#define Color_Nickel           46
#define Color_Rusty            47
#define Color_Silver           48
#define Color_Steel            49
#define Color_Tin              50
#define Color_Titanium         51
#define Color_Tungsten         52
#define Color_Zirconium        53
#define Color_Zinc             54

/* Monsters */
#define Color_Dirt             55
#define Color_Leprous          56
#define Color_Flesh            57  /* Well mine is. Your milage may vary */
#define Color_Disenchanting    58
#define Color_Rotting          59
#define Color_Clay             60
#define Color_Stone            61
#define Color_Fire             62
#define Color_Water            63
#define Color_Earth            64
#define Color_Air              65
#define Color_Frost            66
#define Color_Glowing          67
#define Color_Gelatinous       68
#define Color_Umber            69
#define Color_Crystal          70

/* Mushrooms */
#define Color_Brown            71
#define Color_Ecru             72
#define Color_Furry            73
#define Color_Plaid            74  /* A very tricky Color to define as RGB */
#define Color_Slimy            75
#define Color_Tan              76
#define Color_Wooden           77
#define Color_Wrinkled         78  /* Perhaps try some rayshade textures? */

/* Potions */
#define Color_Icky_Green       79
#define Color_Light_Brown      80
#define Color_Clear            81
#define Color_Azure            82
#define Color_Bubbling         83  /* Animation needed here... */
#define Color_Chartreuse       84
#define Color_Cloudy           85
#define Color_Crimson          86
#define Color_Cyan             87
#define Color_Hazy             88
#define Color_Indigo           89
#define Color_Magenta          90
#define Color_Metallic_Blue    91
#define Color_Metallic_Red     92
#define Color_Metallic_Green   93
#define Color_Metallic_Purple  94
#define Color_Misty            95
#define Color_Pink             96
#define Color_Puce             97
#define Color_Smoky            98
#define Color_Tangerine        99
#define Color_Violet           100
#define Color_Vermilion        101

/* Rocks */
#define Color_Alexandrite      102  /* Emerald Green or Deep Red?? */
#define Color_Amethyst         103
#define Color_Aquamarine       104
#define Color_Azurite          105
#define Color_Beryl            106  /* Emerald or Aquamarine or another? */
#define Color_Bloodstone       107  /* Green with flecks of Red */
#define Color_Calcite          108
#define Color_Carnelian        109
#define Color_Corundum         110  /* Any Color it feels like */
#define Color_Diamond          111
#define Color_Emerald          112
#define Color_Fluorite         113  /* CaF{sub}2 */
#define Color_Garnet           114
#define Color_Granite          115
#define Color_Jade             116
#define Color_Jasper           117  /* Carrot */
#define Color_Lapis_Lazuli     118
#define Color_Magma            119
#define Color_Malachite        120  /* Me! */
#define Color_Marble           121
#define Color_Moonstone        122
#define Color_Onyx             123
#define Color_Pearl            124
#define Color_Quartz           125
#define Color_Quartzite        126
#define Color_Rhodonite        127
#define Color_Ruby             128  /* Both of these are in fact types of */
#define Color_Sapphire         129  /* Corundum (above)...                */
#define Color_Tiger_Eye        130
#define Color_Topaz            131
#define Color_Turquoise        132  /* Did I mention Im doing this by hand? */
#define Color_Zircon           133

/* Treasures */
#define Color_Food             134  /* Pizza-Colored? */
#define Color_Slime            135
#define Color_Leather          136
#define Color_Cord             137  /* This is getting like a Nabakov book */
#define Color_Paper            138
#define Color_Old_Parchment    139  /* "The circus across the park is      */
#define Color_Apple            140  /* too loud."                          */
#define Color_Oil              141
#define Color_Magic_Light      142
#define Color_Mud              143
#define Color_Acid             144
#define Color_Pottery          145
#define Color_Wine             146  /* Villa Maria '86 */
#define Color_Mithril          147

/* Woods */
#define Color_Aspen            148
#define Color_Balsa            149
#define Color_Banyan           150
#define Color_Birch            151
#define Color_Cedar            152
#define Color_Cottonwood       153
#define Color_Cypress          154
#define Color_Dogwood          155
#define Color_Elm              156
#define Color_Eucalyptus       157
#define Color_Hemlock          158
#define Color_Hickory          159
#define Color_Ironwood         160
#define Color_Locust           161
#define Color_Mahogany         162
#define Color_Maple            163
#define Color_Mulberry         164
#define Color_Oak              165
#define Color_Pine             166
#define Color_Redwood          167
#define Color_Rosewood         168
#define Color_Spruce           169
#define Color_Sycamore         170
#define Color_Teak             171
#define Color_Walnut           172

/* Spells/Magic/Breath */
#define Color_Magic_Missile    173
#define Color_Poison_Gas       174
#define Color_Holy_Orb         175

/* New Colors */
#define Color_Lightning        176
#define Color_Deep_Black       177
#define Color_Shadow_And_Flame 178

#define MAX_COLORS             179

/* defined as a special case... */
#define Color_Random           255
#define Color_Opal             Color_Random
#define Color_Multi_Hued       Color_Random
#define Color_Iridescent       Color_Random

/* Defined as previous values */
#define Color_Input                Color_Light_Blue
#define Color_OK                   Color_Green
#define Color_Ok                   Color_OK /* I just know I'll do this... */
#define Color_Attention            Color_Yellow
#define Color_Warning              Color_Red
#define Color_Plain_Text           Color_White
#define Color_Title                Color_White
#define Color_Sub_Title            Color_Light_Grey_High
#define Color_Field                Color_White
#define Color_Non_Applicable       Color_Dark_Grey_High
#define Color_Information          Color_Cream

#define Color_Damned               Color_Red
#define Color_Magik                Color_Green
#define Color_Empty                Color_Dark_Grey_High

#define Color_Inventory_Book       Color_Old_Parchment
#define Color_Inventory_Food       Color_Food
#define Color_Inventory_Potion     Color_Light_Blue
#define Color_Inventory_Scroll     Color_Paper
#define Color_Inventory_Wand       Color_Iron
#define Color_Inventory_Staff      Color_Wooden
#define Color_Inventory_Ring       Color_Gold
#define Color_Inventory_Amulet     Color_Cream
#define Color_Inventory_Armour     Color_Iron
#define Color_Inventory_Weapon     Color_Steel
#define Color_Inventory_Misc       Color_Light_Grey_High
#define Color_Inventory_Enchanted  Color_Green // White

#define Color_Wall                 Color_Medium_Grey_High
#define Color_Floor                Color_Medium_Grey_High

#define Color_Default              Color_White